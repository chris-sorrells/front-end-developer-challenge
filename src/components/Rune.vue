<template>
  <div class="rune-wrapper" :class="{ active }">
    <a
      class="rune"
      :class="[{ active }, { toggleable }, `is-${type}`]"
      @contextmenu.prevent
      @click.left="$emit('activate')"
      @click.right="$emit('deactivate')"
    />
  </div>
</template>

<script>
export default {
  props: {
    active: {
      type: Boolean,
      default: false
    },

    toggleable: {
      type: Boolean,
      default: false
    },

    type: {
      type: String,
      required: true,
      validator(value) {
        return [
          "box",
          "silverware",
          "cake",
          "crown",
          "boat",
          "goggles",
          "lightning",
          "skull"
        ].includes(value);
      }
    }
  }
};
</script>

<style lang="scss">
.rune-wrapper {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 60px;
  height: 60px;
  border: 1px solid #999;
  box-shadow: 0 0 3px 3px rgba(128, 128, 128, 0.25);

  background-image: linear-gradient(
    0deg,
    #6c6c6c,
    #393939 50%,
    #6c6c6c 0,
    #6c6c6c
  );

  &.active {
    background-image: linear-gradient(
      0deg,
      #8fc1f3,
      #243a50 50%,
      #8fc1f3 0,
      #8fc1f3
    );
  }

  .rune {
    display: block;
    background: url("../assets/talent-icons-sprite.png") no-repeat;
    width: 50px;
    height: 50px;
    cursor: pointer;

    &.is-box {
      background-position: 0 -50px;
      &.active,
      &:hover {
        background-position: 0 0;
      }
    }
    &.is-silverware {
      background-position: -50px -50px;
      &.active,
      &:hover {
        background-position: -50px 0;
      }
    }
    &.is-cake {
      background-position: -100px -50px;
      &.active,
      &:hover {
        background-position: -100px 0;
      }
    }
    &.is-crown {
      background-position: -150px -50px;
      &.active,
      &:hover {
        background-position: -150px 0;
      }
    }
    &.is-boat {
      background-position: -200px -50px;
      &.active,
      &:hover {
        background-position: -200px 0;
      }
    }
    &.is-goggles {
      background-position: -250px -50px;
      &.active,
      &:hover {
        background-position: -250px 0;
      }
    }
    &.is-lightning {
      background-position: -300px -50px;
      &.active,
      &:hover {
        background-position: -300px 0;
      }
    }
    &.is-skull {
      background-position: -350px -50px;
      &.active,
      &:hover {
        background-position: -350px 0;
      }
    }
  }
}
</style>
