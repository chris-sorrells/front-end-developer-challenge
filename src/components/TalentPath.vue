<template>
  <div class="talent-path">
    <div v-for="(rune, index) in runes" :key="rune.id" class="rune-container">
      <rune
        :type="rune.type"
        :active="rune.active"
        :toggleable="!runeByIdCanBeToggled(rune.id)"
        @activate="activate(rune.id)"
        @deactivate="deactivate(rune.id)"
      />

      <rune-link v-if="runes.length > index + 1" :active="rune.active" />
    </div>
  </div>
</template>

<script>
import Rune from "@/components/Rune.vue";
import RuneLink from "@/components/RuneLink.vue";
import Debug from "debug";
import { mapGetters } from "vuex";

const debug = Debug("front-end-developer-challenge:RuneContainer.vue");

export default {
  components: {
    Rune,
    RuneLink
  },

  props: {
    runes: {
      type: Array,
      required: true
    }
  },

  computed: {
    ...mapGetters(["runeByIdCanBeToggled"])
  },

  methods: {
    activate(id) {
      return this.$store.dispatch("activate", id).catch(error => {
        debug("Failed to activate rune.", error);
      });
    },

    deactivate(id) {
      return this.$store.dispatch("deactivate", id).catch(error => {
        debug("Failed to deactivate rune.", error);
      });
    }
  }
};
</script>

<style lang="scss">
.talent-path {
  display: flex;

  .rune-container {
    display: flex;
    justify-items: center;
  }
}
</style>
