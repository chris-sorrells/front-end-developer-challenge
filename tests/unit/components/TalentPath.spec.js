import { shallowMount, mount, createLocalVue } from "@vue/test-utils";
import TalentPath from "@/components/TalentPath.vue";
import Rune from "@/components/Rune.vue";
import Vuex from "vuex";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("TalentPath.vue", () => {
  let getters;
  let actions;
  let store;

  let runes = [
    { id: 1, type: "box", active: true, order: 1 },
    { id: 2, type: "skull", active: false, order: 2 }
  ];

  beforeEach(() => {
    getters = {
      runeByIdCanBeToggled: () => () => true
    };

    actions = {
      activate: jest.fn(),
      deactivate: jest.fn()
    };

    store = new Vuex.Store({
      getters,
      actions
    });
  });

  it("renders", () => {
    const wrapper = shallowMount(TalentPath, {
      store,
      localVue,
      propsData: { runes }
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it("dispatches activate action when activating", () => {
    const wrapper = mount(TalentPath, {
      store,
      localVue,
      propsData: { runes }
    });

    wrapper.findComponent(Rune).vm.$emit("activate");
    expect(actions.activate).toBeCalled();
  });

  it("dispatches deactivate action when deactivating", () => {
    const wrapper = mount(TalentPath, {
      store,
      localVue,
      propsData: { runes }
    });

    wrapper.findComponent(Rune).vm.$emit("deactivate");
    expect(actions.deactivate).toBeCalled();
  });

  it("catches activation errors", () => {
    actions.activate = jest.fn().mockRejectedValue("foobar");

    store = new Vuex.Store({
      getters,
      actions
    });

    const wrapper = mount(TalentPath, {
      store,
      localVue,
      propsData: { runes }
    });

    wrapper.findComponent(Rune).vm.$emit("activate");
    expect(actions.activate).toBeCalled();
  });

  it("catches deactivation errors", () => {
    actions.deactivate = jest.fn().mockRejectedValue("foobar");

    store = new Vuex.Store({
      getters,
      actions
    });

    const wrapper = mount(TalentPath, {
      store,
      localVue,
      propsData: { runes }
    });

    wrapper.findComponent(Rune).vm.$emit("deactivate");
    expect(actions.deactivate).toBeCalled();
  });
});
