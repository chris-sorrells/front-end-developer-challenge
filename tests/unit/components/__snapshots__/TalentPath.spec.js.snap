// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TalentPath.vue renders 1`] = `
<div
  class="talent-path"
>
  <div
    class="rune-container"
  >
    <rune-stub
      active="true"
      type="box"
    />
     
    <rune-link-stub
      active="true"
    />
  </div>
  <div
    class="rune-container"
  >
    <rune-stub
      type="skull"
    />
     
    <!---->
  </div>
</div>
`;
